src/cc65/loadexpr.c

```diff
@@ -250,11 +262,15 @@ void LoadExpr (unsigned Flags, struct ExprDesc* Expr)
                 }
 
                 /* And by the width if the field doesn't end on a char or int boundary.
-                ** If it does end on a boundary, then zeros have already been shifted in.
-                ** g_and emits no code if the mask is all ones.
+                ** If it does end on a boundary, then zeros have already been shifted in,
+                ** but we need to clear the high byte for char.  g_and emits no code if the mask
+                ** is all ones.
                 */
-                if (EndBit != CHAR_BITS && EndBit != INT_BITS) {
-                    g_and (F, (0x0001U << Expr->BitWidth) - 1U);
+                if (EndBit == CHAR_BITS) {
+                    /* We need to clear the high byte, since CF_FORCECHAR was set. */
+                    g_and (BitFieldFullWidthFlags | CF_CONST, 0xFF);
```

Comment:
If `BitFieldFullWidthFlags` is supposed to be used in this condition branch, is the operation with `F` on Line 261 wrong?
https://github.com/cc65/cc65/blob/0da27f297394ed7bbf01d7984c545e1c078f6509/src/cc65/loadexpr.c#L253-L262

Reply:
`BitFieldFullWidthFlags` is only needed for the last operation. `g_asr` is the last operation only if `BitOffs == 0 && (EndBit == 8 || EndBit == 16)`, i.e. `!AdjustBitField`.

Reply:
It appears to me this block can be only reached if `(AdjustBitField && !ED_NeedsTest (Expr))`:
`if (AdjustBitField ...)`
https://github.com/cc65/cc65/blob/0da27f297394ed7bbf01d7984c545e1c078f6509/src/cc65/loadexpr.c#L228-L236
`if (AdjustBitField ... && !ED_NeedsTest (Expr))`
https://github.com/cc65/cc65/blob/0da27f297394ed7bbf01d7984c545e1c078f6509/src/cc65/loadexpr.c#L253-L276
And then `F` is no longer used after the last `g_asr` on Line 255 but is still modifed on Line261.

Reply:
Oh, I didn't look at the line numbers correctly in your first answer. yes, that `F |= CF_CHAR` is useless now, but `F` can still be used for the shift. This will all need some tweaking for `long` bit-fields later.

I removed the useless code.

Reply:
I suppose `F == (BitFieldFullWidthFlags | CF_CONST  | CF_FORCECHAR)` when it is used with the `g_asr` here.
`F` could be moved to the `ED_NeedsTest (Expr)` case, and `BitFieldFullWidthFlags` could have `CF_CONST` set from the start so you could just use `BitFieldFullWidthFlags | CF_FORCECHAR` for that `g_asr`, but I am OK with the current code.

Reply:
Yes, for now, but that probably won't be the case in the future with something like `long x : 4`. The type part will probably be different, too.

Reply:
Uh, I am fine with the latest revision https://github.com/cc65/cc65/pull/1153/commits/5f8e50a454d167b98b709d0021a3d64a5b715608.

Sorry I haven't made it clear.